Feature public headers

evedon opened the pull request:
### Description

Separated platform, usb, drivers, events, and rtos internal APIs from public APIs

* Moved source files under `source` subdirs
* Moved header files included by public headers under `internal` subdirs
* Added Doxygen comments for documenting internal and public APIs
* Removed source code from header files in order to remove include pre-processor directives
* Used explicit include header files instead of implicit inclusions via third-party header files

Refactored the `usb` directory

* The contents of the `usb` directory were moved to appropriate locations and the `usb` directory removed
  * Public usb headers moved under `drivers`
  * Internal usb headers moved under `drivers/internal/`
  * Source code moved under `drivers/source/usb/device`
  * `usb/device/hal/`  moved under `hal/usb/`
  * `usb/device/USBPhy/` moved under `hal/usb/`
  * Merged `usb/device/targets/` into `targets/`

### Pull request type

    [ ] Fix
    [X] Refactor
    [ ] Target update
    [ ] Functionality change
    [X] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@bulislaw 

### Release Notes

This will break user code that was using an internal API using a prefixed path, for example `#include "foo/bar.h"` instead of `#include "bar.h"`.

hugueskamba committed:
Platform: Separate internal APIs from public APIs

Also includes:
* rename `mbed_sleep_manager.c` to `mbed_power_mgmt.c` to match its
  header file
* create Doxygen groups for public and internal APIs
* use relative path to include header files where inconsistent
* update references to internal APIs throughout libraries
* update the copyright year for all modified files

hugueskamba committed:
Drivers/Events/RTOS Public and internal APIs cleanup (#10955)

Separate drivers, events, and rtos internal APIs from public APIs.

* Move source files to source subdirs
* Move internal headers to internal subdirs
* Add Doxygen comments for documenting internal and public APIs
* Remove source code from header files in order to remove include pre-processor directives
that included header files not directly used by said header files
* Explicitly include header files instead of implicit inclusions via third-party header files.

Release Notes

This will break user code that was using an internal API as the internal header files have been moved.
This will only break if the user was including the header file using a namespace (i.e #include "foo/bar.h" instead of #include "bar.h"

evedon committed:
Fixing include files due to latest rebase

gpsimenos committed:
USB public APIs cleanup (#11034)

The contents of the usb directory were moved to appropriate locations and the usb directory removed.

* Public USB headers moved under drivers/
* Internal USB headers moved under drivers/internal/
* USB Source code moved under drivers/source/usb/
* Moved usb/device/hal/ under hal/usb/
* Moved usb/device/USBPhy/ under hal/usb/
* Merged usb/device/targets/ into targets/
* Separated public and private USB API documentation under Doxygen groups drivers-public-api and drivers-internal-api.

bulislaw commented on drivers/source/DigitalIn.cpp:
```diff
@@ -15,14 +15,17 @@
  */
 
-#include "drivers/AnalogIn.h"
+#include "drivers/DigitalIn.h"
```
This diff looks weird, could you confirm it's just git's thing and not a real issue.

hugueskamba replied:
I can confirm this is a git diff issue. The history of the file is accurate.
https://github.com/ARMmbed/mbed-os/commits/feature-public-headers/drivers/source/DigitalIn.cpp
It is a newly added file

More info [here](https://github.com/ARMmbed/mbed-os/pull/10955/#issuecomment-510387821)

bulislaw reviewed (changes requested):
Could we change the doxygen groups structure a bit: having two main groups public-api and internal-api without split per directory seems like something that is easy to reference to and also should be easier to browse. Then having each of the main directories as it's own group (drivers, platforms, etc). Also it would be good to group related APIs together. So having SPI group for spi, spislave and qspi; i2c for i2c and i2cslave; usb for all usb classes (without group per class); timers for all normal and low power timer related classes; uart for all uart and serial; gpio and analog. There probably some other groupings that would make sense so please add them if you spot something.

Is there a way we can check how much we are affecting the character limit as pointed out by Seppo?

Before that goes in could someone have a look at hand generated doxygen and see if there's something obviously broken? Also would be good to run all the nightiles and examples, Qinghao will be able to help here.

bulislaw commented on events/internal/equeue.h:
```diff
@@ -26,11 +24,14 @@ extern "C" {
 #include <stddef.h>
 #include <stdint.h>
 
+/** \ingroup events */
```
I think that should be a public header. I know it's sort of underlying concept, but in some situations (eg bare metal) people may not want to use C++. We also need to do something smart about the both README files, I would say merge them into one. Starting with the C++ description and then introduce concept of underlying C APIs and that they can be used in bare metal c only projects.

bulislaw replied:
@geky fyi

geky replied:
Well, opinions aside, this just made all the subtrees quite a bit harder to maintain :(

Are we doing the same for littlefs?

0Grit replied:
I can't even imagine sub-trees being any harder to maintain than they already were when I last tried to use sub trees for dependency management.

geky replied:
Oh really? I'm actually a big fan of subtrees :) Though as I understand the quality of the command has changed quite a bit over the last half-decade.

It works really well at separating version control from other programmer debates such as monorepo/multirepo, build/dependency systems.

It can get complicated, but one of the best things it has going for it is that, all the effort is on the developer making the subtree merge. After that the user never needs to know (unlike submodules) and even other developers can work on merged directory as though it was part of the repo.

Every so often I'll come back through and create another subtree merge pr to bring the code up to date with what's outside the tree (though I've been slow because of littlefs v2 stuff, sorry about that): https://github.com/ARMmbed/mbed-os/pull/7713

Subtrees really seem like a realization of the idea of truly distributed version control. Because of it repos can exist as copies in a bunch of different places with their own quirks and patches.

geky replied:
After thinking about this, I think it's time that equeue and Mbed OS split and go their separate ways.

In general, the feedback I've gotten on equeue in Mbed OS has been negative. I think at this point it's overstayed its welcome.

Not that Mbed OS should go without an event queue, just that Mbed OS's event queue should be revamped and get a new owner (@kjbracey-arm, nanostack-eventloop? 🙂).

equeue originally started as a standalone, OS-agnostic project over here: https://github.com/geky/equeue. I still maintain it, though it's relatively quiet. The goal was to have a small, reliable, and useful standalone tool for scheduling on microcontrollers, because, from my limited experience, this is all you need 90% of the time. And by being OS-agnostic, you can write an application that is more or less portable across different environments.

You may notice this doesn't really match the idea behind Mbed OS, like, at all. And I think that's creating some conflicting design decisions.

It's in equeue's best interest to keep the API as minimal as possible, as this reduces the surface area for bugs and quirky interactions. This is why I push back so hard on new features in an effort to stop feature-creep and keep the scope limited. But I'm just creating problems if your only platform is Mbed OS and you're trying to solve some Mbed OS-centric issues:
- event priorities https://github.com/ARMmbed/mbed-os/issues/10263
- static allocation https://github.com/ARMmbed/mbed-os/issues/9172
- cancel to return success https://github.com/ARMmbed/mbed-os/pull/10750

Going the other way, Mbed OS doesn't need a number of things in equeue:
- FreeRTOS/Window/etc support
- Backgrounding for when you only have hardware timers
- Chaining makes less sense when you have threads 

So my conclusion is that these two projects should diverge.

A graceful way to do this would be to rename equeue -> mbed_events or something similar in Mbed OS so it matches the name of C++ API. @bulislaw I think you're right we should merge the READMEs and I think that extends to the directory structure as well. Basically abolish the separation that currently exists because equeue is also a standalone repo. I know it's not great, but a name change would be very helpful as this way equeue can also be brought in as a separate library without name conflicts.

Then, once time permits, we can introduce nanostack-eventloop and mbed-events to each other and see what happens.

That's my proposal, if we're moving around a lot of files it would be awfully convenient to do this now. I can give you a sed script or add a commit if you'd like.

Thoughts?

@bulislaw, @kjbracey-arm, @pan-, @c1728p9, @evedon

40Grit replied:
Though I have a preference for single threaded event queing over multi-threading...

Your statements above feel right.

Mbed OS needs an event queing system that can be more flexible in it's interactions with other components of the OS and more actively maintained.

I'd say turn your statement above into it's own. Issue and go for it.

pan- replied:
@geky Let's not throw the baby out with the bathwater. There's some things the `EventQueue` does better than other event systems present in mbed-os tree :
* It's not a singleton and allows you to have several queue coexisting in a system. 
* Allocation system works great with C++ lambda
* Allocation is per queue; not global.  

Consolidation of a single event system would be great to reduce code size, maintenance effort and bug surface but, first, we need to gather requirements from all these systems (`EventQueue`, `nano stack event loop`, `WSF event queue`, ...) to make sure we're going in the right direction before jumping on a solution.

----

> You may notice this doesn't really match the idea behind Mbed OS, like, at all. And I think that's creating some conflicting design decisions.

How so ? Queuing event will **always** be relevant and if it can run on x86 it can help for integration and unit testing as well.

evedon replied:
This is a useful discussion but I am very reluctant to add scope to this already huge PR. So except for @bulislaw original comment on whether equeue should be regarded as a public or internal API, I'd like everything else to be dealt with separately, specifically: the README cleanup and the potential renaming of equeue to mbed_events.

geky replied:
@evedon, good point. I'll move this to a separate issue.

My vote for equeue public vs internal is to leave it as internal for now. Are we worried this will break code?

geky replied:
> Let's not throw the baby out with the bathwater. There's some things the EventQueue does better than other event systems present in mbed-os tree :

To be clear I'm not suggesting we remove EventQueue. Or even remove any features yet.

Just that we rename equeue -> mbed_events (or similar), remove separate between equeue and EventQueue where it makes sense, and let the projects naturally diverge as their features grow/shrink to match their requirements.

> How so ? Queuing event will always be relevant and if it can run on x86 it can help for integration and unit testing as well.

I was more referring specifically to the portability/OS-agnostic aspect of equeue.

You're right x86 support has been useful for testing, but IMO x86 testing should somehow be provided for all of Mbed OS. It's the only way to keep this shindig scalable.

FreeRTOS and Windows support have already been removed from the Mbed OS version because it makes less sense here. And there have been several cases of the terminology differences between equeue and the RTOS confusing users.

bulislaw replied:
I agree :D So how about for this PR we:
* Make queue public, as we need C API for bare metal builds
* Merge the readmes so it all still makes sense 
* Do not rename anything public so that we don't break compatibility

Going forward:
* @kjbracey-arm is looking at getting all the requirements from all the components so we can finaly have one queue to rule them all (mandatory https://xkcd.com/927/)
* @donatieng @pan- are looking at 2 priorities for the queue and at static allocations

For mbed 6 we will try to force the unification and possible changes to the existing APIs.

I'll catch up with y'all at some point to discuss the plans further.

Does it sound like a plan?

0Grit replied:
👍 
@bulislaw @kjbracey-arm @donatieng @pan  
I am looking forward to the discussion of this topic further in the related public issue/issues

geky replied:
I'm a bit confused. My conclusion was that there shouldn't be a single queue. Is there plans to maintain a version outside of mbed-os?

> @donatieng @pan- are looking at 2 priorities for the queue and at static allocations

FYI, I am also looking at static allocations, (though intentionally not priorities). A part that plan was to break API compatibility to introduce better error codes. Let me know if you want me to create a PR here as well though.

bulislaw reviewed (commented):
We also should deal a bit better with the `rtos` directory, there's source which matches our current model, but there's also `TARGET_CORTEX` which probably should go under the `sources`. I don't have strong opinion what to do with the files from `TARGET_CORTEX`. And similiar for TARGET_CORTEX_M under `platform/`

hugueskamba committed:
Fix Ethernet driver build (#11096)

Include the missing header file inclusion to find the added
MBED_DEPRECATED macro.

hugueskamba committed:
Change Doxygen groups structure, splitting first by Public/Internal (#11105)

* Change Doxygen groups structure, splitting first by Public/Internal

This commit also does the following:
* groups the documentation of related API
* moves `events/internal/equeue.h` to `events/equeue.h`
* merges `events/source/README.md` to `events/README.md`

gpsimenos committed:
Fix CI for branch feature-public-headers (#11093)

* Fix rtos include path in NRFCordioHCIDriver
* Flatten USB driver directory structure
* Add missing include for us_ticker
* Add more missing includes for us_ticker
* Fix mbed_hal_fpga_ci_test_shield/uart test
* Fix bare-metal build
* Fix Watchdog UNITTEST
* Fix Mbed OS 2 build for Public/Internal headers relocating

hugueskamba committed:
Address comments on workaround for Mbed OS 2 CI build after Public  (#11114)


* Modify compilation API to provide a list of paths to exclude from the build.
* `_exclude_files_from_build` becomes a static method
* Replace ternary expression with simple  `if/else` statement
* Make unit test case for dirs exclusion independent of system files

bulislaw commented on drivers/BusIn.h:
```diff
@@ -23,12 +23,19 @@
 namespace mbed {
-/** \addtogroup drivers */
+/** \defgroup drivers-public-api-bus Bus
+ * \ingroup drivers-public-api
```
That should also be GPIO

bulislaw commented on drivers/BusInOut.h:
```diff
@@ -22,15 +22,18 @@
-/** \addtogroup drivers */
+/**
+ * \defgroup drivers_BusInOut BusInOut class
+ * \ingroup drivers-public-api-bus
```
Another one for GPIO group.

bulislaw commented on drivers/BusOut.h:
```diff
@@ -22,10 +22,13 @@
 namespace mbed {
-/** \addtogroup drivers */
+/**
+ * \defgroup drivers_BusOut BusOut class
```
And another for GPIO

bulislaw commented on drivers/CAN.h:
```diff
@@ -27,12 +27,19 @@
+ */
+
+/**
+ * \defgroup drivers_CANMessage CANMessage class
```
Why are we defining a group for every single driver? If it's only one does it need it's own group?

hugueskamba replied:
This is to create an index for the class `CANMessage` within the group `CAN` which itself is in the group `Drivers` found in the group `Public API` of the Doxygen document generated.
The Doxygen `CAN` group has more than one class: `CANMessage` and `CAN`.

There are no group for every single driver. For example there is no group for `Watchdog` or `ResetReason`. Drivers have been grouped only if it made sense to do so.

Have a look at the Doxygen document to see if there is something you see as not quite right.

bulislaw commented on drivers/Ethernet.h:
```diff
@@ -54,9 +60,13 @@ namespace mbed {
- * @ingroup drivers
  */
-class Ethernet : private NonCopyable<Ethernet> {
+class
```
Could we extend the doxygen to sort of repeat it so it shows up in the documentation as well as warning if someone uses it.

bulislaw commented on drivers/InterruptIn.h:
```diff
@@ -29,7 +29,15 @@
 namespace mbed {
-/** \addtogroup drivers */
+/** \defgroup drivers-public-api-interrupt Interrupt
+ * \ingroup drivers-public-api
```
That should probably also go into GPIO.

bulislaw commented on drivers/InterruptManager.h:
```diff
@@ -18,13 +18,17 @@
 
 namespace mbed {
-/** \addtogroup drivers */
+/**
```
GPIO as well.

bulislaw commented on drivers/PortIn.h:
```diff
@@ -22,10 +22,13 @@
 
 namespace mbed {
-/** \addtogroup drivers */
+/**
```
GPIO

hugueskamba replied:
It already is. See line 29.

bulislaw commented on drivers/PortInOut.h:
```diff
@@ -22,15 +22,17 @@
 
 namespace mbed {
-/** \addtogroup drivers */
+/**
```
GPIO

hugueskamba replied:
It already is. See line 29.

bulislaw commented on drivers/PortOut.h:
```diff
@@ -22,10 +22,14 @@
 
 namespace mbed {
-/** \addtogroup drivers */
+/**
```
GPIO

hugueskamba replied:
It already is. See line 29.

bulislaw commented on drivers/SerialWireOutput.h:
```diff
@@ -18,29 +18,26 @@
 
 namespace mbed {
+/**
+ * \defgroup drivers_SerialWireOutput SerialWireOutput class
```
That's not really UART, SWO is a debug protocol: https://en.wikipedia.org/wiki/JTAG#Serial_Wire_Debug we can create a debug group.

hugueskamba replied:
Is there something else to group it with? What else will be in the debug group? If there is nothing a group will not be created.

bulislaw commented on drivers/internal/PolledQueue.h:
```diff
@@ -18,18 +18,26 @@
 #include "LinkedList.h"
 namespace events {
-/** \addtogroup events */
+/** \addtogroup events-internal-api Events
```
That's still USB internal I think?

bulislaw commented on drivers/internal/Task.h:
```diff
@@ -19,12 +19,15 @@
 
 namespace events {
-/** \addtogroup events */
+/** \addtogroup events-internal-api
```
Is that USB rather than event?

bulislaw commented on drivers/internal/TaskBase.h:
```diff
@@ -27,15 +27,18 @@ class Semaphore;
 class TaskQueue;
 
+/**
+ * \defgroup events_TaskBase TaskBase class
```
Is that USB?

bulislaw commented on drivers/internal/TaskQueue.h:
```diff
@@ -18,21 +18,23 @@
 
+/**
+ * \defgroup events_TaskQueue TaskQueue class
+ * \ingroup events-internal-api
```
Is that USB?

bulislaw commented on events/mbed_events.h:
```diff
@@ -20,24 +16,21 @@
 
 #include "events/mbed_shared_queues.h"
 
+/** \ingroup mbed-os-internal */
```
Why do we have that here?

bulislaw commented on features/device_key/source/DeviceKey.h:
```diff
@@ -32,7 +32,7 @@
 
 namespace mbed {
-/** \addtogroup drivers */
+/** \addtogroup device-key Device Key */
```
drivers-security would be probably good group for now

bulislaw commented on platform/internal/CallChain.h:
```diff
@@ -28,7 +28,8 @@ namespace mbed {
 class CallChainLink;
 
-/** \addtogroup platform */
+/** \ingroup mbed-os-internal */
```
I don't think it's internal.

evedon replied:
Do you mean that it should be public? 
At the start of the Epic I asked @kjbracey-arm for a classification of platform APIs betwen public and internal and CallChain was classified as "Private / deprecated (remove?)"
I see this API only used in `InterruptManager.h`

bulislaw reviewed (changes requested):
Partner files that move should be reviewed by the parnters teams. I can tell the NXP are moved in wrong place. There's no source files in the directory where they are being moved to. Looking at which NXP partner platforms are touched you should be able to figure out where to push the files.

hugueskamba committed:
Upstream PR #11073 review request changes (#11135)

* Modify Doxygen grouping of `drivers` Public/Internal APIs
* Correct classification of `mbed_events.h`
* Amend name of Doxygen group containing Device Key API
* Classify `CallChain.h` as public API and relocate file
* Remove Doxygen group from `equeue_platform.h` as it has no Doxygen compliant documentation
* Move USB target specific code back to `usb/device/targets`

hugueskamba committed:
Mark Ethernet.h methods as Deprecated in Doxygen documentation (#11146)

bulislaw reviewed (approved):
Looks good, can we run the automated example to verify we don't break our own use cases? @jamesbeyond
